Policy: Fix auth policies with proxy redirects

jrajahalme opened the pull request:
When covered entries with lower proxy port priority have different explicit auth types, the proxy port (and priority) must be propagated to those entries instead of deleting them. Same applies in reverse, when adding an explicit auth entry the proxy port and priority must be propagated from a covering entry with higher priority proxy port, if any.

Keep the non-auth policy handling simple by branching to authPreferredInsert early, and take care of deny & higher priority proxy port logic in the same loops with the auth propagation. This required changing 'CoveringKeysWithSameID' and 'SubsetKeysWithSameID' to also iterate keys equal to the 'newKey', so that a deny key would not be overridden by an allow key, for example.

Add a new unit test to prevent future regressions.

Fixes: #36056

```release-note
Auth policy is properly maintained also when covered by proxy redirects.
```

jrajahalme committed:
policy: Propagate higher precedence proxy ports to covered auth entries

When covered entries with lower proxy port priority have different
explicit auth types, the proxy port (and priority) must be propagated to
those entries instead of deleting them. Same applies in reverse, when
adding an explicit auth entry the proxy port and priority must be
propagated from a covering entry with higher priority proxy port, if any.

Keep the non-auth policy handling simple by branching to
authPreferredInsert early, and take care of deny & higher priority proxy
port logic in the same loops with the auth propagation. This required
changing 'CoveringKeysWithSameID' and 'SubsetKeysWithSameID' to also
iterate keys equal to the 'newKey', so that a deny key would not be
overridden by an allow key, for example.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

doniacld reviewed (approved):
Nice work! 
nit: mispelling`propagaed` in your second commit message.

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -746,17 +752,28 @@ func (ms *mapState) insertWithChanges(newKey Key, newEntry mapStateEntry, featur
+// overrideProxyPortForAuth sets the proxy port and priority of 'v' to that of 'newKey', saving the
+// old entry in 'changes'.
+// Returns 'true' if changes were made.
+func (ms *mapState) overrideProxyPortForAuth(newEntry mapStateEntry, k Key, v mapStateEntry, changes ChangeState) bool {
```
adding a placeholder comment just to pause merge while I give this a quick grok

squeed commented on pkg/policy/mapstate_test.go:
```diff
@@ -1341,6 +1342,38 @@ func TestMapState_AccumulateMapChanges(t *testing.T) {
+		name:      "test-5d - proxy port propagation from the most specific covering key with a proxy port - reverse",
+		args: []args{
+			{cs: csBar, adds: []int{43}, port: 80, proto: 6, prefix: 16, authReq: AuthTypeSpire.AsExplicitRequirement()},
+			{cs: csFoo, adds: []int{43}, port: 80, proto: 6, prefix: 12, redirect: true},
```
want to add another entry with longer prefix, just to make sure it's carried down correctly?

jrajahalme replied:
Yes, and found issues while doing so!

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -725,6 +725,12 @@ func (ms *mapState) insertWithChanges(newKey Key, newEntry mapStateEntry, featur
 		}
 	} else {
+		// authPreferredInsert takes care for precedence and auth
+		if features.contains(authRules) {
```
Does it really make sense to skip the precedence logic separately if auth is enabled? I don't love the duplicated logic between auth and non-auth paths.

I'd rather see the BroaderOrEqualKeys iteration pass to some kind of `mergeParent()` function that can either merge auth from above or prevent insertion of the child element in the event the parent has higher precedence.

jrajahalme replied:
Without this the main business logic here became much more complicated and much harder to read. That's why I decided to keep the added complexity in `authPreferredInsert()`, even when it now needs to also implement the proxy port precedence logic. There was no clean way to separate proxy port precedence logic from auth propagation logic any more, and given that the non-auth code path is actually used in production IMO this is the right choice.

jrajahalme replied:
Also, when we take care of auth and proxy port precedence in the same loop we only need to iterate through covering and covered entries once.

jrajahalme replied:
Auth propagation also uses a bit more expensive iteration that iterates in LPM order that is not needed for the non-auth case. So really `authPreferredInsert()` implements the generic logic for adding an allow entry now, and the non-auth variant in `insertWithChanges` is a faster implementation for the most common case.

jrajahalme committed:
policy: Skip recording delete if added on the same round of changes

Only add a delete into ChangeState if the key was not added on the same
round of 'changes'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add new proxy port propagation tests

Proxy port with a higher precedence should be propagated to a covered
entry instead of deleting it, if the covered entry has a different
explicit auth type than the covering proxy redirect entry.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -777,30 +792,66 @@ func (ms *mapState) overrideAuthRequirement(newEntry mapStateEntry, k Key, v map
+			//   not have existed, and
+			// - since the new entry has explicit auth it does not need to be
+			//   derived.
+			break
```
What if there are two covering entries with proxy ports? Then it would not be safe to bail out here, we would need to "inherit" twice?

jrajahalme replied:
In that case the most specific covering entry would have the highest priority proxy port in it (either from the L4Filter, or "inherited"), so there is no need to look at the less specific covering entries.

squeed replied:
Aha, missed that `CoveringKeysWithSameID` iterates most-to-least-specific.
